### ISSUE TITLE
page view count

### ISSUE THREAD

**Issue by steviepubliclab** [publiclab/plots2#1166]
### What happened just before the problem occurred

It looks like the view count on the public lab wikis pages aren't working right now. Hard to tell if the same is going on for the Research Notes, I've seen some with really low counts that should be high (such as this one https://publiclab.org/notes/lenagd/12-20-2016/citizen-science-investigations-aka-common-legal-issues-when-using-community-sourced-data), and some that look ok (such as this one: https://publiclab.org/notes/sanchittechnogeek/01-03-2017/translating-project-files-using-transifex) maybe the count was just reset? Would be great to get them back if we can!

**Comment by jywarren** [comment]
I believe @yukiisbored just implemented a new counter system, but perhaps the old view counts weren't added into the new system. @yukiisbored @ananyo2012 is there a way we can run a new migration to fetch the view counts from the old system, perhaps with the now-deprecated code:

```ruby
self.drupal_node_counter.totalcount
```

and insert `impressionist` counts with null IP addresses?

**Comment by ananyo2012** [comment]
@jywarren Uh quite a similar thing that happened back when I tried to integrate `friendly_id` for updating slugs. It didn't save the old view count. Looking for a possible solution.

**Comment by jywarren** [comment]
I believe the existing `self.drupal_node_counter.totalcount` values should still be there, so it should be a matter of testing in the rails console if it's possible to increment the impressionist counter manually without needing to supply unique IP addresses; once solved, we might simply add a few lines to a new migration file to do so automatically in a new PR.

**Comment by ananyo2012** [comment]
There are two ways to solve this problem -- by setting a initial count to the `totalcount` field to the counter. Or we can store unique ip addresses in the `impressions` table which seems to be a tedious one. The first one seems to be an easier solution.

**Comment by Pk13055** [comment]
@ananyo2012 I'm willing to work on this issue if you reach a stand on which method to use.

**Comment by ananyo2012** [comment]
@Pk13055 I suggest the first one. Let @jywarren reply what should be the previous view counts.

**Comment by jywarren** [comment]
Hi, Ananyo - can you elaborate on your first suggestion? I think
`drupal_node_counter.totalcount` should still have the original counts --
how would we get impressionist to recognize that, or could we run a batch
script (in a Rails Migration) to add that number to Impressionist? I think
we should look at the Impressionist docs to see how this can be done.

**Comment by ananyo2012** [comment]
@jywarren No the `totalcount` got updated and recalculated as the `totacount` field is added as the counter_cache column. See [this line](https://github.com/publiclab/plots2/blob/master/app/models/drupal_node_counter.rb#L7).

Briefing the technique how impressionist calculates the `totacount` -- Whenever a controller request is made a new record is created in the `impressions` table ( see [this migration](https://github.com/publiclab/plots2/blob/master/db/migrate/20161230142222_create_impressions_table.rb) ) and the `ip_address` gets stored. The counter_cache which is the `totalcount` field here gets recalculated and is incremented only if the request is made by an unique ip address.

**Comment by jywarren** [comment]
Oh, that's unfortunate :-(

Well, we have staging data from not too long ago. And I can ask Sebastian
how accessible last weeks database dump is.

Please keep an eye out for changes that can irreversibly destroy data -- we
were lucky that this time it's only pageviews.

Ananyo - can you tell how we are able to add our recovery data to
totalcount again without Impressionist recalculating it? I can look into
generating a per-node listing (CSV or something) with `nid,count` for each
node from a recovery database.

**Comment by ananyo2012** [comment]
@jywarren Looking for a possible solution to set a initial count to the `totalcount` field.

Actually if we look closely the page view count didn't reflect the unique view count right from the beginning. So impressionist was supposed to revamp the system completely and form a new view count for each content such as research note,wikis and questions. I was aware of that. But the thing I missed was the earlier view count had no record of the requests made. So impressionist reset it to zero due to that.

**Comment by jywarren** [comment]
Hey, no worries, that's why we do backups, and I missed this too. Great,
ping me if you find something, hopefully we can get a fix in pretty soon.
Thank you!

**Comment by Pk13055** [comment]
@jywarren @ananyo2012 Do tell me if you reach a stand on this; I'll try to get a fix in ASAP.

**Comment by calaway** [comment]
Hello, my friend @tgisg and I were just looking through your issues and were hoping we might be able to replicate this bug. When we loaded the page below two times from two different IP addresses we confirmed that both views __were__ counted. I can't say if the counter was reset, as postulated in the issue description, but it seems to be working from this point.

https://publiclab.org/notes/lenagd/12-20-2016/citizen-science-investigations-aka-common-legal-issues-when-using-community-sourced-data

**Comment by jywarren** [comment]
Hello -- yes, apologies -- the issue is currently that we're looking to recover the count from just before this was merged; thank you! The feature now works properly, it's just the existing counts we'd like to preserve.

If you're looking for a place to get involved, do check out:

https://github.com/publiclab/plots2/projects/2?fullscreen=true

and also our "on-deck" not-quite fully formatted first-timers issues, at:

https://github.com/publiclab/plots2/labels/fto-candidate

Thank you so much for your interest!

**Comment by jywarren** [comment]
I looked through the docs for Impressionist (https://github.com/charlotte-ruby/impressionist) and I wasn't able to figure out a good way to insert views without inserting impressionist records. 

I'm considering possibly creating a new column on `node_counter` or `node` which stores the legacy counts, and returning the sum of that and `impressionist`'s count... not ideal, but not sure what else to do here...

**Comment by jywarren** [comment]
Also put in a request to Sebastian to recover the old counts from a database backup.

**Comment by jywarren** [comment]
So [this line](https://github.com/publiclab/plots2/pull/1153/files#diff-506c305b074506147da2da7a9d9d6dcfR479) in `drupal_node.rb` would become:

```ruby
def totalcount
      DrupalNodeCounter.new({:nid => self.id}).save if self.drupal_node_counter.nil?
     self.drupal_node_counter.totalcount + self.legacy_pageviews
end
```

Make sense? And we could re-enter the legacy counts manually.

**Comment by yukiisbored** [comment]
Yeah, we can create a new column on `drupal_node_counter` and add them with the actual page counter. Sorry that i caused some troubles for you guys, I should've thought about it and ensure legacy values are still being added with the new one.

**Comment by yukiisbored** [comment]
What we can do is to grab the old view counter from the backups, create a new column in `drupal_node_counter` to hold the old view counts, add them on `totalcount`with the new value, and set them accordingly.

**Comment by yukiisbored** [comment]
We probably can hack up a script, but I'm not that familliar with ActiveRecord nor Rails, but i'll look into it and see what I can do.

**Comment by jywarren** [comment]
I'm working on a script, assuming a CSV like:

```
nid,totalcount
1,100
2,100
3,100
...
```

So my script is:

```ruby
CSV.foreach('counts.csv') do |row|
  DrupalNode.find(row[0]).update_attributes(legacy_pageviews: row[1])
end
```

I'm working on the migration now too.

**Comment by jywarren** [comment]
OK - i've also gone ahead and merged `node_counter` into `node`, and we don't sort by view count much anymore (preferring "most recently updated") but there are still a few, which'll act a bit weird since they won't see the legacy counts:

```
grep -r "views DESC" app/
app/models/tag.rb:              .order("node.views DESC")
app/controllers/questions_controller.rb:    @questions = @questions.order('views DESC')
app/controllers/wiki_controller.rb:                       .order("views DESC")
app/controllers/notes_controller.rb:                       .order("views DESC")
```

Pushing up a PR now, it'll need some close checking.

**Comment by jywarren** [comment]
OK - submitted in #180
